### PR TITLE
Add Events shortcode

### LINKED ISSUE #881: Next Gen WordCamp Experiment: Accepting applications
# Purpose
This issue serves as a place to discuss technical details in supporting "[The Next Generation of WordCamps](https://make.wordpress.org/community/2023/05/08/the-next-generation-of-wordcamps/)", focusing on accepting applications.

# Features/Tasks
1. Update the [Become an Organizer](https://central.wordcamp.org/become-an-organizer/) page to lead users into the relevant form
1. Create a page with the application form 
1. Add the ability to approve/reject
## Update the [Become an Organizer](https://central.wordcamp.org/become-an-organizer/) page to lead users into the relevant form
Ideally, I think we end up with a dedicated landing page for these events, but for the interim we can probably tie into the existing WordCamp application flow to direct applicants.

**Page**
Screenshot 2023-05-23 at 9 44 13 AM

## Create a page with the application form 
We currently have [shortcodes](https://github.com/WordPress/wordcamp.org/tree/production/public_html/wp-content/plugins/wcpt/views/applications) that render html forms. We should probably stick with that since there isn't a Gutenberg block for this yet. We'll _lose_ the ability for admins to modify the form since fields are likely to change during the beta, but that tradeoff seems fine for now.

## Acceptance Criteria

1. A new page with a form container information from [this comment](https://github.com/WordPress/wordcamp.org/issues/881#issuecomment-1843798832)
2. Upon submission the form should create a `wordcamp` post type.

### PR DESCRIPTION
**Pull request by pkevan**
Allows the creation of Events shortcode to gather application data, rather than using PD form.

Testing this is somewhat tricky due to it saving data on Central, and there being no test environment to check.

I've done some limited testing without saving or sending notifications to slack, but will be best to try this after a merge and deploy, with some mock events to make sure it's all working correctly.

Fixes #881

Testing instructions:

Add the shortcode `[events-organizer-application]` to a page (can be draft) - complete the form, and the data should save as a `WordCamps` post type with all data saved. Only some fields are mapped into the post type.

### CODE REVIEW

**Comment by iandunn** [comment]
> no test environment to check

What's the reason it can't be tested in the Docker env?

**Comment by pkevan** [comment]
> > no test environment to check
> 
> What's the reason it can't be tested in the Docker env?

I don't have docker currently setup - I'll have a look at this later.

**Comment by pkevan** [comment]
Finally got docker working and installed (not without it's challenges). Had to alter the code to make it work locally (but not on sandbox):

wcpt-events/class-events-application.php needed changing to:
```
L144 $statues   = \WordCamp_Loader::get_post_statuses();
```

Shortcode renders function, and saves event in `WordCamp` post type in the status of `needs vetting`, data is saved into fields as specified in the original issue with the majority of data being saved but not mapped since the Community team goes through the data to check.

**Comment by iandunn** [comment]
> Finally got docker working and installed (not without it's challenges). 

I did a fresh provision last week and fixed a few problems in #1187 and #1188. Those were the only problems I ran into, though. If you still remember what you ran into, do you mind creating an issue for it?

> Had to alter the code to make it work locally (but not on sandbox):
> L144 $statues   = \WordCamp_Loader::get_post_statuses();

🤔 Wouldn't it need to be that everywhere? I don't see a `get_post_statuses` method in that class, only in `WordCamp_Loader`. There is a `get_post_statuses()` function from Core, but IIRC PHP won't attempt to resolve `self::foo()` to `\foo()` if it doesn't find `self::foo()`.

File: public_html/wp-content/plugins/wcpt/views/applications/events/shortcode-application.php [hunk 1]
```diff
@@ -0,0 +1,504 @@
...
+
+							<div>
+								<label for="q_add2">Address Line 2</label>
+								<br/>
+								<input type="text" maxlength="100" class="b" name="q_add2" id="q_add2"
+									   value=""/>
+							</div>
+
+							<div>
+								<label for="q_city">City</label>
```
Comment (StevenDufresne) [inline public_html/wp-content/plugins/wcpt/views/applications/events/shortcode-application.php:119]:
```suggestion
								<label for="q_city">City <span class="PDF_mand">*</span></label>
```

File: public_html/wp-content/plugins/wcpt/views/applications/events/shortcode-application.php [hunk 2]
```diff
@@ -0,0 +1,504 @@
...
+						</div>
+					</div>
+				</div>
+
+				<div class="PDF_question" id="pd-question-18">
+					<div class="qContent">
+						<div class="qText">
+							Your event URL will be: https://events.wordpress.org/[location]/[year]/[event-name]
+							e.g. https://events.wordpress.org/narnia/2024/communityday
+							What would you like to be the [event-name] at the end of your event URL?
```
Comment (StevenDufresne) [inline public_html/wp-content/plugins/wcpt/views/applications/events/shortcode-application.php:433]: At the very least "What would ..." should be on its own line. 

It's very wordy though, do we need to pass this effort along to the user in this manner? @hellojuliarose

**Comment by StevenDufresne** [comment]
On submitting this form locally, I received this error:

```
Notice: Trying to get property 'ID' of non-object in /usr/src/public_html/wp-content/plugins/wcpt/wcpt-events/class-events-application.php on line 143 

Fatal error: Uncaught Error: Call to undefined method WordPress_Community\Applications\Events_Application::get_post_statuses() in /usr/src/public_html/wp-content/plugins/wcpt/wcpt-events/class-events-application.php:144 

Stack trace: #0 /usr/src/public_html/wp-content/plugins/wcpt/wcpt-event/class-event-application.php(135): WordPress_Community\Applications\Events_Application->create_post(Array) 
#1 /usr/src/public_html/wp-content/plugins/wcpt/wcpt-event/class-event-application.php(67): WordPress_Community\Applications\Event_Application->submit_application(Array) 
#2 /usr/src/public_html/mu/wp-includes/shortcodes.php(433): WordPress_Community\Applications\Event_Application->render_application_shortcode('', '', 'events-organize...') 
#3 [internal function]: do_shortcode_tag(Array) 
#4 /usr/src/public_html/mu/wp-includes/shortcodes.php(273): preg_replace_callback('/\\[(\\[?)(events...', 'do_shortcode_ta...', '[events-organiz...') 
#5 /usr/src/public_html/mu/wp-includes/class-wp-hook.php(324): do_shortcode('[events-organiz...') 
#6 in /usr/src/public_html/wp-content/plugins/wcpt/wcpt-events/class-events-application.php on line 144
```

It seems legit, `self::get_post_statuses()` does appear missing.

**Comment by pkevan** [comment]
>  If you still remember what you ran into, do you mind creating an issue for it?

Sure - there were several :) I'll go through my notes. I also fixed some documentation in https://github.com/WordPress/wordcamp.org/pull/1198

**Comment by pkevan** [comment]
Going to merge this and deploy ready for next week.

**Comment by pkevan** [comment]
noting that a couple of feedback items didn't get address in this PR, but will add them in another one.

**Pull request merged**